### ISSUE TITLE
(OpenWrt) How to automatically start the v2ray instance on system startup?

### ISSUE THREAD

**Issue by lss4** [v2rayA/v2rayA#468]
### Version, approach of installation, operating system

1. What version of v2rayA are you using?

1.5.6.2

2. What approach of installation do you install V2rayA?

Installed to FW via OpenWrt ImageBuilder (`make image`).
Device is Zyxel NBG6817 (ipq806x).

3. What is your operating system?

OpenWrt latest snapshot (r19194 at the time of writing)

### Description

On OpenWrt, while the v2raya itself can be configured to start on boot, I cannot find any option to make it automatically start v2ray. I always have to manually go to the page upon router startup/restart and start the v2ray instance from there.

I've seen several other issues about this, but it doesn't seem to be very related to OpenWrt usage.

### Logs：

 ```
2022/03/26 02:39:48.321  [A] [main.go:25]  network is connected
2022/03/26 02:39:48.378  [A] [main.go:27]  V2Ray binary is /usr/bin/v2ray
2022/03/26 02:39:48.378  [A] [main.go:27]  V2Ray asset directory is /run/user/0/v2raya
2022/03/26 02:39:48.378  [A] [main.go:27]  v2rayA working directory is /
2022/03/26 02:39:48.378  [A] [main.go:27]  v2rayA configuration directory is /etc/v2raya
2022/03/26 02:39:48.378  [A] [main.go:27]  Golang: go1.17.8
2022/03/26 02:39:48.378  [A] [main.go:27]  OS: linux
2022/03/26 02:39:48.379  [A] [main.go:27]  Arch: arm
2022/03/26 02:39:48.379  [A] [main.go:27]  Lite: false
2022/03/26 02:39:48.381  [A] [main.go:27]  Version: 1.5.6.2
2022/03/26 02:39:48.381  [A] [main.go:27]  Starting...
2022/03/26 02:39:48.418  [A] [index.go:117]  v2rayA is listening at http://127.0.0.1:2017
... (a few more listening addresses afterwards and that is it, until I manually start the server from the webpage)
 ```

I need to manually go to the page on port 2017 to manually start v2ray, to change it from Ready to Running state. I don't see any apparent errors, either.

### v2rayA setting and related screenshot

I don't find the option to automatically start the v2ray server configuration on system startup anywhere.

**Comment by mzz2017** [comment]
@lss4 What is your log level

**Comment by lss4** [comment]
> @lss4 What is your log level

What log level? v2raya or v2ray-core?

There's no option to change it from the UI at all, and the v2ray-core log level appears to be "info" according to `config.json`.

**Comment by mzz2017** [comment]
@lss4 It is weird if there is no error. Did you shutdown your PC by accident?

**Comment by lss4** [comment]
> @lss4 It is weird if there is no error. Did you shutdown your PC by accident?

Ugh. This isn't a PC. It's my router (Zyxel NBG6817).

**Comment by mzz2017** [comment]
@lss4 https://github.com/v2rayA/v2rayA/actions/runs/2043426216

Replace your v2raya binary with it; it will print a log to show if the running status was normal the last time v2raya exited.

**Comment by lss4** [comment]
Okay... it seems on OpenWrt there are race conditions going on, preventing v2raya from starting v2ray-core on initial startup.

```
2022/03/26 01:56:43.420  [E] [main.go:28]  failed to start v2ray-core: Please sync datetime first. Your datetime is 2022/03/26 01:56, and the correct datetime is 2022/03/26 04:47
```
When my router boots up its system time is not defined. I think that (01:56) was the time when I built the system image.

It can take a while before the system time gets synced, and until then, v2ray-core will refuse to launch with the error message above. I'm currently using VLESS which is not bound by system timestamp, though.

Additionally, if I issue `/etc/init.d/v2raya restart` later on, I get this message which seems to indicate that v2raya did not properly close the previous v2ray-core on shutdown. v2raya shows Ready instead of Running state as it does not manage the still-open v2ray-core instance launched by the previous v2raya instance. I need to manually close the previous v2ray-core process from shell in order to be able to launch another v2ray-core instance from the v2raya page.

```
2022/03/26 04:59:56.323  [E] [main.go:28]  failed to start v2ray-core: port is occupied by v2ray(6753): 20170
```

I cannot trivially experiment with your v2ray binary too much, as this requires me to manually write another v2raya binary into the overlay partition, which is very low on space, so it was tested with a symlink. Using a symlink this way may impair the auto startup of v2raya itself (not v2ray-core), but as this is not a standard configuration for my use case, I will not dig into this deeper.

**Comment by lss4** [comment]
I've reverted to the original binary I installed into my firmware. After cleaning up the previous v2ray process, this time v2raya successfully started a new v2ray-core instance when it started. I haven't rebooted my router this time, though.

So the issue seems to be the following:
- `v2raya` is starting too early. This might also be due to the way with OpenWrt's `init` system works. Don't know if it's possible to delay its startup a bit longer, so the time sync would likely have done its job by the time v2raya starts.
- `v2raya` does not handle the shutdown and restart (#380) of the `v2ray` process. The `v2ray` process launched by previous v2raya needs to be killed by hand should v2raya be restarted.

PS: It seems v2ray-core has a high memory requirement as this message can be seen when it starts. This is not a big issue at the moment, though.
```
2022/03/26 04:49:52.237  [I] [process.go:67]  low memory: 307MiB, set V2RAY_CONF_GEOLOADER=memconservative
```

**Comment by mzz2017** [comment]
@lss4 Maybe you can use sleep 3 && v2raya to replace the original command

**Comment by lss4** [comment]
> @lss4 Maybe you can use sleep 3 && v2raya to replace the original command

Okay... I found [the line](https://github.com/v2rayA/v2rayA/blob/4a0d47cf537a4f19d89518165b4870879888a0fd/service/core/v2ray/processManager.go#L49) where the "sync datetime first" error is.

I think this time check should only be used if the device loads initial timestamp from a RTC powered by a secondary power source (such as a lithium cell). For devices that are instead designed to always load a default timestamp from somewhere (such as zero, or image build timestamp) on startup, this will cause issues.

**Comment by lss4** [comment]
A bump on this.

It seems when v2raya is being active (v2ray running) for a prolonged amount of time (like about 7 days) v2raya could be killed with no apparent messages from system or v2raya itself, either an OOM occurred or a bug happened from the inside. The underlying v2ray process remained up the moment v2raya was restarted by the system, so v2raya could not restart it at that point, until it got killed at a later point, at that point I could then start v2ray from v2raya page.

Maybe v2raya should save the PID of the underlying v2ray somewhere and attempt to check the PID status when it starts, as well as attempt to kill it when it gracefully closes. In case v2raya finds the old v2ray process remained active when it starts, it should force-kill it before starting the new one as otherwise it won't be able to start it because the ports are still being taken by the old one.

**Comment by lss4** [comment]
Update on this again. Apparently under some circumstances v2rayA can bring up the default proxy on OpenWrt startup.

However, most likely due to memory or internal issues, v2ray is not stable and from time to time it gets killed with no apparent warnings, forcing me to manually go to the v2rayA page to reopen the proxy.

Found some excerpts from a messy, netcat-captured router system log (typed here manually):

```
panic: runtime error: index out of range [0] with length 0

goroutine 172106573 [running]:
  github.com/v2rayA/v2rayA/core/specialMode/infra.(*handle).handlePacket(0x286a198, {0xf77300, 0x2b31ef0}, {0x2a2ad60, 0x6}, 0x2cb5260, 0x2cc0018)
  github.com/v2rayA/v2rayA/core/specialMode/infra/handle.go:188 +0xf60
created by github.com/v2rayA/v2rayA/core/specialMode/infra.(*DnsSupervisor).Run
  github.com/v2rayA/v2rayA/core/specialMode/infra/supervisor.go:114 +0x358
```

PS: It seems the v2rayA I'm running on the router is a bit outdated (I'm currently on 1.5.6.2). Will try this on 1.5.7 or onwards when I build a new image for the router (not now). Not sure if the related lines have been touched on later versions... but still, I think running this on a router is never going to be a good idea.

**Comment by mzz2017** [comment]
@lss4 turn off the supervisor please
